docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+# Microbenchmark Design Guidelines
+
+## Table of Contents
+
+- [Mindset](#Mindset)
+  - [Benchmarks are not Unit Tests](#Benchmarks-are-not-Unit-Tests)
+  - [Benchmarks are Immutable](#Benchmarks-are-Immutable)
+- [BenchmarkDotNet](#BenchmarkDotNet)
+- [Setup](#Setup)
+  - [GlobalSetup](#GlobalSetup)
+  - [IterationSetup](#IterationSetup)
+  - [OperationsPerInvoke](#OperationsPerInvoke)
+- [Test Cases](#Test-Cases)
+  - [Code Paths](#Code-Paths)
+    - [Array.Reverse](#Array.Reverse)
+    - [Buffer.CopyMemory](#Buffer.CopyMemory)
+  - [Always the same input data](#Always-the-same-input-data)
+  - [BenchmarkDotNet](#BenchmarkDotNet)
+    - [Arguments](#Arguments)
+    - [Params](#Params)
+    - [ArgumentsSource](#ArgumentsSource)
+    - [Generic benchmarks](#Generic-benchmarks)
+- [Best Practices](#Best-Practices)
+  - [Single Responsibility Principle](#Single-Responsibility-Principle)
+  - [No Side-Effects](#No-Side-Effects)
+  - [Dead Code Elimination](#Dead-Code-Elimination)
+  - [Loops](#Loops)
+  - [Method inlining](#Method-Inlining)
+- [Summary](#Summary)
+
+## Mindset
+
+Writing Benchmarks is much different than writing Unit Tests. So before you start coding, you need to change your mindset first.
+
+### Benchmarks are not Unit Tests
+
+When writing Unit Tests, we ideally want to test all methods and properties of the given type. We also test both the happy and unhappy paths. The result of every Unit Test run is a single value: passed or failed.
+
+Benchmarks are different. First of all, the result of a benchmark run is never a single value. It's a whole distribution, described with values like mean, standard deviation, min, max and so on. To get a meaningful distribution, the benchmark has to be executed many, many times. **This takes a lot of time**. With the current [recommended settings](https://github.com/dotnet/performance/blob/51d8f8483b139bb1edde97f917fa436671693f6f/src/harness/BenchmarkDotNet.Extensions/RecommendedConfig.cs#L17-L20) used in this repository, it takes on average six seconds to run a single benchmark. 
+The public surface of .NET Standard 2.0 API has `36 519` methods. If we had 1 benchmark for every public method, it would take two and a half days to run the benchmarks.
```

Comment:
While the precision is nice we aren't going to keep this up to date and I don't want PRs "fixing" it :)

```suggestion
The public surface of .NET Standard 2.0 API has tens of thousands of methods. If we had 1 benchmark for every public method, it would take two and a half days to run the benchmarks.
```